src/Data/Functor/Foldable/TH.hs

```diff
@@ -70,12 +74,59 @@ import Paths_recursion_schemes (version)
 -- as we don't try to do better than
 -- <https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#deriving-functor-instances GHC's DeriveFunctor>.
 --
-makeBaseFunctor :: Name -> DecsQ
-makeBaseFunctor = makeBaseFunctorWith baseRules
+-- Some doctests:
+--
+-- >>> data Expr a = Lit a | Add (Expr a) (Expr a) | Expr a :* [Expr a]
+-- >>> ; makeBaseFunctor ''Expr
+--
+-- >>> :t AddF
+-- AddF :: r -> r -> ExprF a r
+--
+-- >>> data Rose f a = Rose a (f (Rose f a))
+-- >>> ; makeBaseFunctor [d| instance Functor f => Recursive (Rose f a) |]
+--
+-- >>> :t RoseF
+-- RoseF :: a -> f r -> RoseF f a r
+--
+-- >>> let rose = Rose 1 (Just (Rose 2 (Just (Rose 3 Nothing))))
+-- >>> cata (\(RoseF x f) -> x + maybe 0 id f) rose
+-- 6
+--
+class MakeBaseFunctor a where
+    makeBaseFunctor :: a -> DecsQ
+
+instance MakeBaseFunctor a => MakeBaseFunctor [a] where
+    makeBaseFunctor a = fmap concat (T.mapM makeBaseFunctor a)
+
+instance MakeBaseFunctor a => MakeBaseFunctor (Q a) where
+    makeBaseFunctor a = makeBaseFunctor =<< a
+
+instance MakeBaseFunctor Name where
+    makeBaseFunctor = makeBaseFunctorWith baseRules
+
+instance MakeBaseFunctor Dec where
```

Comment:
Add some Haddocks here to explain that the only `Dec`s that are permitted are instance declarations.

Reply:
Added